test/e2e/run-openrpc-api-test-coverage.ts

```diff
@@ -0,0 +1,644 @@
+  let result = await driver.executeScript(`return window['${generatedKey}'];`);
+
+  while (result === null) {
+    // Continue polling if result is not set
+    await driver.delay(50);
+    result = await driver.executeScript(`return window['${generatedKey}'];`);
+  }
+
+  // clear the result
+  await driver.executeScript(`delete window['${generatedKey}'];`);
+
+  return result;
+};
+
+const createDriverTransport = (driver: Driver) => {
+  return async (
+    _: string,
+    method: string,
+    params: any[] | Record<string, any>,
+  ) => {
+    const generatedKey = uuid();
+    // don't wait for executeScript to finish window.ethereum promise
+    // we need this because if we wait for the promise to resolve it
+    // will hang in selenium since it can only do one thing at a time.
+    // the workaround is to put the response on window.asyncResult and poll for it.
+    driver.executeScript(
+      ([m, p, g]: any) => {
+        window[g] = null;
+        window.ethereum
+          .request({ method: m, params: p })
+          .then((r: any) => {
+            window[g] = { result: r };
+          })
+          .catch((e: any) => {
+            window[g] = {
+              error: {
+                code: e.code,
+                message: e.message,
+                data: e.data,
+              },
+            };
+          });
+      },
+      method,
+      params,
+      generatedKey,
+    );
+    const response = await pollForResult(driver, generatedKey);
+    return response;
+  };
+};
+
+interface ConfirmationsRejectRuleOptions {
+  driver: Driver;
+  only: string[];
+}
+// this rule makes sure that all confirmation requests are rejected.
+// it also validates that the JSON-RPC response is an error with
+// error code 4001 (user rejected request)
+class ConfirmationsRejectRule implements Rule {
+  private driver: any;
+  private only: string[];
+  private rejectButtonInsteadOfCancel: string[];
+  private requiresEthAccountsPermission: string[];
+
+  constructor(options: ConfirmationsRejectRuleOptions) {
+    this.driver = options.driver;
+    this.only = options.only;
+    this.rejectButtonInsteadOfCancel = [
+      'personal_sign',
+      'eth_signTypedData_v4',
+    ];
+    this.requiresEthAccountsPermission = [
+      'personal_sign',
+      'eth_signTypedData_v4',
+      'eth_getEncryptionPublicKey',
+    ];
+  }
+
+  getTitle() {
+    return 'Confirmations Rejection Rule';
+  }
+
+  async beforeRequest(_: any, call: Call) {
+    try {
+      if (this.requiresEthAccountsPermission.includes(call.methodName)) {
+        const requestPermissionsRequest = JSON.stringify({
+          jsonrpc: '2.0',
+          method: 'wallet_requestPermissions',
+          params: [{ eth_accounts: {} }],
+        });
+
+        await this.driver.executeScript(
+          `window.ethereum.request(${requestPermissionsRequest})`,
+        );
+        const screenshot = await this.driver.driver.takeScreenshot();
```

Comment:
are there actually two nested `driver` properties like this 🫠 ?

Reply:
Maybe we should pull it out to be a method on our driver class...

not a blocker but